pygfunction/borefield.py

```diff
@@ -45,10 +47,9 @@ def __init__(
             self, H: npt.ArrayLike, D: npt.ArrayLike, r_b: npt.ArrayLike,
             x: npt.ArrayLike, y: npt.ArrayLike, tilt: npt.ArrayLike = 0.,
             orientation: npt.ArrayLike = 0.):
-        # Convert x and y coordinates to arrays
-        x = np.atleast_1d(x)
-        y = np.atleast_1d(y)
-        self.nBoreholes = np.maximum(len(x), len(y))
+        self.nBoreholes = np.max(
+            [len(np.atleast_1d(var))
+             for var in (H, D, r_b, x, y, tilt, orientation)])
```

Comment:
Did you consider an assertion for when `len(x) != len(y)`?

What about enforcing `H`, `D`, `r_b`, `tilt` and `orientation` to match `len(<var>) == len(y)` or `len(<var>) == 1`?

Reply:
An assertion would certainly be important here. The goal was to mimic numpy array broadcasting rules and allow for example `x = [0., 5., 10.], y = 0.` if 3 boreholes are on the same row.

Reply:
The `Borehole` class can represent both a segment of a borehole or the borehole itself. Likewise, the collection of parameters in `Borefield` can represent segments or boreholes. With that in mind, segments along a vertical borehole share the same coordinates (`x`, `y`) and only require `H` and `D` arrays at instantiation.

One way to approach this would be to use `b = np.broadcast(H, D, r_b, x, y, tilt, orientation)` to make sure that the inputs are broadcastable to a common shape with `b.nd <= 1`. The number of boreholes is then `nBoreholes = b.size`.

Reply:
Feel free to make use of any ideas contained in commit https://github.com/j-c-cook/pygfunction/commit/bd6e46792064c64b76c54cf8752b839c1d249c23. I've enjoyed the back and forth and little bit of collaboration here. I'll likely drop out of discussion now.

Reply:
I adapted your changes in #322. I removed the try / except clauses since `numpy.broadcast` raises an error if the arguments cannot be broadcast.